auth/providers/azure/graph/graph.go

```diff
@@ -242,38 +206,57 @@ func (u *UserInfo) Name() string {
 	return getterName
 }
 
-// New returns a new UserInfo object
-func New(clientID, clientSecret, tenantName string, useGroupUID bool, aadEndpoint, msgraphHost string) (*UserInfo, error) {
-	graphEndpoint := "https://" + msgraphHost + "/"
-	baseAPIURL, _ := url.Parse(graphEndpoint + "v1.0")
-	loginURL := aadEndpoint + "%s/oauth2/v2.0/token"
-
-	parsedLogin, err := url.Parse(fmt.Sprintf(loginURL, tenantName))
-	if err != nil {
-		return nil, err
-	}
+// newUserInfo returns a UserInfo object
+func newUserInfo(tokenRefresher TokenRefresher, graphURL *url.URL, useGroupUID bool) (*UserInfo, error) {
 	u := &UserInfo{
 		client: http.DefaultClient,
 		headers: http.Header{
 			"Content-Type": []string{"application/json"},
 		},
-		apiURL:        baseAPIURL,
-		loginURL:      parsedLogin,
-		clientID:      clientID,
-		clientSecret:  clientSecret,
-		groupsPerCall: expandedGroupsPerCall,
-		useGroupUID:   useGroupUID,
-		defaultScope:  graphEndpoint + ".default", // This requests that a token use all of its default scopes
+		apiURL:         graphURL,
+		groupsPerCall:  expandedGroupsPerCall,
+		useGroupUID:    useGroupUID,
+		tokenRefresher: tokenRefresher,
 	}
 
 	return u, nil
 }
 
+// New returns a new UserInfo object
+func New(clientID, clientSecret, tenantID string, useGroupUID bool, aadEndpoint, msgraphHost string) (*UserInfo, error) {
+	graphEndpoint := "https://" + msgraphHost + "/"
+	graphURL, _ := url.Parse(graphEndpoint + "v1.0")
+
+	tokenRefresher := NewClientCredentialTokenRefresher(clientID, clientSecret,
+		fmt.Sprintf("%s%s/oauth2/v2.0/token", aadEndpoint, tenantID),
+		fmt.Sprintf("https://%s/.default", msgraphHost))
+
+	return newUserInfo(tokenRefresher, graphURL, useGroupUID)
+}
+
+// NewWithOBO returns a new UserInfo object
+func NewWithOBO(clientID, clientSecret, tenantID string, aadEndpoint, msgraphHost string) (*UserInfo, error) {
+	graphEndpoint := "https://" + msgraphHost + "/"
+	graphURL, _ := url.Parse(graphEndpoint + "v1.0")
+
+	tokenRefresher := NewOBOTokenRefresher(clientID, clientSecret,
+		fmt.Sprintf("%s%s/oauth2/v2.0/token", aadEndpoint, tenantID),
+		fmt.Sprintf("https://%s/.default", msgraphHost))
+
+	return newUserInfo(tokenRefresher, graphURL, true)
+}
+
+// NewWithAKS returns a new UserInfo object used in AKS
+func NewWithAKS(tokenURL, tenantID, msgraphHost string) (*UserInfo, error) {
+	graphEndpoint := "https://" + msgraphHost + "/"
+	graphURL, _ := url.Parse(graphEndpoint + "v1.0")
+
+	tokenRefresher := NewAKSTokenRefresher(tokenURL, tenantID)
```

Comment:
OBOTokenRefresher has 'scope' passed in, but NewAKSTokenRefresher doesn't?
Does it mean the service NewAKSTokenRefresher calls would internally have a hardcoded scope (resource in AAD v1)?
Shall we make this 2 consistent? either OBOTokenRefresher also hardcode graphendpoint as scope, or NewAKSTokenRefresher also accept a scope?
What do you think

Reply:
IMO, OBOTokenRefresher provides full flexibility to caller because it is unmanaged so it requires user input. For AKSTokenRefresher, it will be a black box to caller so that it doesn't require any input other than just an endpoint to call